rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -430,69 +418,29 @@ rmw_create_publisher(
-    allocator->allocate(sizeof(rmw_zenoh_cpp::rmw_publisher_data_t), allocator->state));
+  auto node_data = context_impl->get_node_data(node);
   RMW_CHECK_FOR_NULL_WITH_MSG(
-    publisher_data,
-    "failed to allocate memory for publisher data",
+    node_data,
+    "NodeData not found.",
     return nullptr);
-  auto free_publisher_data = rcpputils::make_scope_exit(
-    [publisher_data, allocator]() {
-      allocator->deallocate(publisher_data, allocator->state);
-    });
 
-  RMW_TRY_PLACEMENT_NEW(
-    publisher_data, publisher_data, return nullptr,
-    rmw_zenoh_cpp::rmw_publisher_data_t);
-  auto destruct_publisher_data = rcpputils::make_scope_exit(
-    [publisher_data]() {
-      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
-        publisher_data->~rmw_publisher_data_t(), rmw_zenoh_cpp::rmw_publisher_data_t);
-    });
-
-  generate_random_gid(publisher_data->pub_gid);
-
-  // Adapt any 'best available' QoS options
-  publisher_data->adapted_qos_profile = *qos_profile;
-  rmw_ret_t ret = rmw_zenoh_cpp::QoS::get().best_available_qos(
-    node, topic_name, &publisher_data->adapted_qos_profile, rmw_get_subscriptions_info_by_topic);
-  if (RMW_RET_OK != ret) {
+  if (!node_data->create_pub_data(
+    rmw_publisher,
+    context_impl->session(),
+    context_impl->get_next_entity_id(),
+    topic_name,
+    type_support,
+    qos_profile)) {
+    // Error already handled.
     return nullptr;
   }
 
-  publisher_data->typesupport_identifier = type_support->typesupport_identifier;
-  publisher_data->type_hash = type_support->get_type_hash_func(type_support);
-  publisher_data->type_support_impl = type_support->data;
-  auto callbacks = static_cast<const message_type_support_callbacks_t *>(type_support->data);
-  publisher_data->type_support = static_cast<rmw_zenoh_cpp::MessageTypeSupport *>(
-    allocator->allocate(sizeof(rmw_zenoh_cpp::MessageTypeSupport), allocator->state));
-  RMW_CHECK_FOR_NULL_WITH_MSG(
-    publisher_data->type_support,
-    "Failed to allocate MessageTypeSupport",
-    return nullptr);
-  auto free_type_support = rcpputils::make_scope_exit(
-    [publisher_data, allocator]() {
-      allocator->deallocate(publisher_data->type_support, allocator->state);
-    });
-
-  RMW_TRY_PLACEMENT_NEW(
-    publisher_data->type_support,
-    publisher_data->type_support,
-    return nullptr,
-    rmw_zenoh_cpp::MessageTypeSupport, callbacks);
-  auto destruct_msg_type_support = rcpputils::make_scope_exit(
-    [publisher_data]() {
-      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
-        publisher_data->type_support->~MessageTypeSupport(),
-        rmw_zenoh_cpp::MessageTypeSupport);
-    });
-
-  publisher_data->context = node->context;
-  rmw_publisher->data = publisher_data;
+  // Store type erased node in rmw_publisher->data so that the
+  // PublisherData can be safely accessed.
+  rmw_publisher->data = reinterpret_cast<void *>(const_cast<rmw_node_t *>(node));
```

Comment:
I had to rely on a `reinterpret_cast<const_cast<>>` to be able to store the `const rmw_node_t * node` available to `rmw_create_publisher()` within `rmw_publisher->data` as the `data` field is `void *` and not `const void *`. 

This is a bit messy. Another alternative I considered was storing the `rmw_context_impl_s *` ptr here and then including an `unordered_map` in this class to obtain the `NodeData` for a given `rmw_publisher_t` but this involves extra bookkeeping that we need to do when we create and destroy a publisher. Open to other suggestions!

Reply:
It's a bit ugly, but I think it is OK.  At least, I think it is the best we can do given our constraints.